### PR TITLE
Updates to Spack in standard Nextflow

### PR DESCRIPTION
**Pull request by marcodelapierre**
This Update PR brings a number of changes to the Spack integration in plain Nextflow (NOT Wave):
  1. [user experience] `spack.checksum` instead of `spack.noChecksum` (also in docs)
  2. [robustness] add `spack env view enable` after activate for spack yaml (also in docs)
  3. [robustness] add `spack config add concretizer` in cmds for both spack yaml and pkgs
  4. [functionality] support for archs via Spack target option (also in docs)
      * now implemented as generalised `process.arch` directive, with support for both Docker and Spack notations

I have updated also code unit tests, and performed runtime tests.

The key update is number 4. 
[EDIT]  This is now compliant with https://github.com/nextflow-io/nextflow/issues/3701.

AFTER this PR: add  arch support in Wave

### CODE REVIEW

**Comment by marcodelapierre** [comment]
@pditommaso have a look, including implementation for point 4. above

**Comment by marcodelapierre** [comment]
@pditommaso can you have a look at this one?
I think it is better to consolidate arch support here (Spack with plain Nextflow) before adding it into the Wave Spack PR.

**Comment by pditommaso** [comment]
My only concern about merging this PR is that it contains an implementation of `arch` directive that does not match https://github.com/nextflow-io/nextflow/issues/3701, and this will include in the upcoming stable release scheduled by next week. 

What about keeping `arch` outside of this PR and just use `x86_64` for the Spack build?

**Comment by marcodelapierre** [comment]
oh no, I was not clear, sorry Paolo.

can you review the PR up to this point? then do not merge, after your approval I will carry on work to implement #3701 in the PR before your next approval and merging.

File: modules/nextflow/src/main/groovy/nextflow/spack/SpackCache.groovy
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/spack/SpackCache.groovy:351]: A conditional expression can make this more concise e.g. 

```
spackEnvArch = arch ? spackEnv + "_" + arch : spackEnv
```

File: modules/nextflow/src/main/groovy/nextflow/spack/SpackCache.groovy
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/spack/SpackCache.groovy:359]:
```suggestion
            def msg = spack found local environment `$spackEnv`
            if( arch ) msg += " and arch `$arch`" 
            log.trace "spack found local environment `$spackEnv` for arch `$arch`"

```

File: modules/nextflow/src/main/groovy/nextflow/spack/SpackCache.groovy
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/spack/SpackCache.groovy:375]: As previous comment

**Review by pditommaso — CHANGES_REQUESTED** [review]
Left some minor comments

File: modules/nextflow/src/main/groovy/nextflow/spack/SpackCache.groovy
Comment (marcodelapierre) [reply modules/nextflow/src/main/groovy/nextflow/spack/SpackCache.groovy:351]: adding it

File: modules/nextflow/src/main/groovy/nextflow/spack/SpackCache.groovy
Comment (marcodelapierre) [reply modules/nextflow/src/main/groovy/nextflow/spack/SpackCache.groovy:359]: adding it

File: modules/nextflow/src/main/groovy/nextflow/spack/SpackCache.groovy
Comment (marcodelapierre) [reply modules/nextflow/src/main/groovy/nextflow/spack/SpackCache.groovy:375]: adding it

**Comment by marcodelapierre** [comment]
.. done!

so @pditommaso , if you give me green lights, I will proceed to implement the arch specifications as per #3701

**Comment by marcodelapierre** [comment]
This will address #3701

**Comment by marcodelapierre** [comment]
@pditommaso  @bentsherman  I have finally finished work on this one, see updated description on top;
The key, last addition of this week is the generalised implementation of the `process.arch` directive, which supports both Docker and Spack notations.

Very keen to get your feedback on this one! :-)

File: docs/spack.rst
Comment (bentsherman) [inline docs/spack.rst:86]:
```suggestion
Spack is able to infer the local CPU microarchitecture and optimise the build accordingly.
```

File: modules/nextflow/src/main/groovy/nextflow/processor/Arch.groovy
Comment (bentsherman) [inline modules/nextflow/src/main/groovy/nextflow/processor/Arch.groovy:32]: Let's name this class `Architecture`, even if the directive is `arch`

File: modules/nextflow/src/main/groovy/nextflow/processor/TaskConfig.groovy
Comment (bentsherman) [inline modules/nextflow/src/main/groovy/nextflow/processor/TaskConfig.groovy:380]: Similarly, I think this should be `getArchitecture()`

File: modules/nextflow/src/main/groovy/nextflow/spack/SpackCache.groovy
Comment (marcodelapierre) [reply modules/nextflow/src/main/groovy/nextflow/spack/SpackCache.groovy]: tested, then reverted, because Gradle build was failing

Comment (bentsherman) [reply modules/nextflow/src/main/groovy/nextflow/spack/SpackCache.groovy:346]: Hmm, you probably have to do `"${spackEnv}_${arch}".toString()`

Comment (marcodelapierre) [reply modules/nextflow/src/main/groovy/nextflow/spack/SpackCache.groovy]: thanks done!

File: modules/nextflow/src/main/groovy/nextflow/processor/Arch.groovy
Comment (marcodelapierre) [reply modules/nextflow/src/main/groovy/nextflow/processor/Arch.groovy]: done!

File: modules/nextflow/src/main/groovy/nextflow/processor/TaskConfig.groovy
Comment (marcodelapierre) [reply modules/nextflow/src/main/groovy/nextflow/processor/TaskConfig.groovy]: done!

**Comment by marcodelapierre** [comment]
Thanks @bentsherman , I have addressed all of your current feedback.

File: modules/nextflow/src/main/groovy/nextflow/processor/Architecture.groovy
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/processor/Architecture.groovy:56]: Since the class is marked `@CompileStatic`, this is not needed here

File: modules/nextflow/src/main/groovy/nextflow/processor/Architecture.groovy
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/processor/Architecture.groovy:89]: Better to throw the exception here instead returning an invalid string. I guess you made for testing, but still better to split the test in two and validate ok value and invalid values separately

File: modules/nextflow/src/main/groovy/nextflow/processor/TaskConfig.groovy
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/processor/TaskConfig.groovy:383]:
```suggestion
        if( value instanceof CharSequence )
            return new Architecture(value.toString())
```

Otherwise `GString` (groovy strings) are not matched

File: modules/nextflow/src/main/groovy/nextflow/script/ProcessConfig.groovy
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/script/ProcessConfig.groovy:954]: Worth adding a test in `ProcessConfigTest`

**Review by pditommaso — CHANGES_REQUESTED** [review]
Nice job, we are almost there! Just left a few comments

File: modules/nextflow/src/main/groovy/nextflow/processor/Architecture.groovy
Comment (marcodelapierre) [reply modules/nextflow/src/main/groovy/nextflow/processor/Architecture.groovy:56]: done!

File: modules/nextflow/src/main/groovy/nextflow/processor/Architecture.groovy
Comment (marcodelapierre) [reply modules/nextflow/src/main/groovy/nextflow/processor/Architecture.groovy:89]: Actually, I did it because I did not want to pass `res.name` as an additional argument to the function.

But I agree with your point, so ... done!

File: modules/nextflow/src/main/groovy/nextflow/processor/TaskConfig.groovy
Comment (marcodelapierre) [reply modules/nextflow/src/main/groovy/nextflow/processor/TaskConfig.groovy:383]: ok, I trust you in full...done!

File: modules/nextflow/src/main/groovy/nextflow/script/ProcessConfig.groovy
Comment (marcodelapierre) [reply modules/nextflow/src/main/groovy/nextflow/script/ProcessConfig.groovy:954]: done!

**Comment by marcodelapierre** [comment]
OK, @pditommaso  I have tackled your feedback, many thanks for that!

**Comment by marcodelapierre** [comment]
@pditommaso the CI error is a GitHub rate limit issue

**Comment by marcodelapierre** [comment]
and now CI failing with

```
docker: Error response from daemon: unauthorized: repository 'docker.io/pditommaso/my-secret-container:latest' unauthorized (401).
```

**Comment by pditommaso** [comment]
it should happen because builds on a fork do not inherit the secrets.

**Review by pditommaso — APPROVED** [review]
let's merge this. Thanks Marco!

**Comment by marcodelapierre** [comment]
Fantastic! Thank you!

Tomorrow I will jump straight away on the Wave one, I want to see it done!

**Comment by pditommaso** [comment]
Let's do it!

**Pull request merged**
